# printing


In [4]:
print("hello world")
print("dbhwjkdw")

hello world
dbhwjkdw


# variable= a container for a value (string,integer,float,boolean)


In [5]:
first="simon"

print(first)

simon


In [8]:
print("hello",first)
a="food"
print("i like",a)

hello simon
i like food


In [10]:
age=25
print("you are ",age," years old")

you are  25  years old


In [11]:
price=10.99
print(f"the price is${price}")

the price is$10.99


In [12]:
# boolean
isstudent= True
print(f"are you a student: {isstudent}")

are you a student: True


In [13]:
if isstudent:
    print("ebfjhfb")
else:
    print("dwuegytwe")

ebfjhfb


# typecasting = the process of converting a variable from one data type to another str(), int(),float(),bool()

In [17]:
name="simon j kurian"
age=20
gpa=3.2
is_student=True
print(type(name),type(age),type(gpa),type(is_student))


<class 'str'> <class 'int'> <class 'float'> <class 'bool'>


In [22]:
gpa=int(gpa)
print(gpa)
print(type(gpa))

age=float(age)
print(age)

age=str(age)
print(type(age))


3
<class 'int'>
20.0
<class 'str'>


In [23]:
age+=1

TypeError: can only concatenate str (not "int") to str

In [24]:
print(age+hello)

NameError: name 'hello' is not defined

In [26]:
print(age+"hello")

20.0hello


In [27]:
print(age,"30")

20.0 30


In [28]:
name=bool(name)
print(name)

True


In [29]:
a=""
a=bool(a)
print(a)

False


# input() = a function that prompts the user to enter data returns the entered data as STRING

In [39]:
input()

 wedfgh


'wedfgh'

In [41]:
input("what is your name?")

what is your name? stephin


'stephin'

In [32]:
input(djvhwjh)

NameError: name 'djvhwjh' is not defined

In [34]:
name=input("what is your name?")
print(f"hello {name}")

what is your name? simon


hello simon


In [35]:
age=input("how old are you?")
age=age+1
print(age)   # so only string is accepted

how old are you? simon


TypeError: can only concatenate str (not "int") to str

In [40]:
# age=int(input()) 
age=int(input("how old are you?"))
age=age+1
print("happy birthday")
print(age)

how old are you? 25


happy birthday
26


# exercise 1

In [41]:
length=input("enter the length")
b=input("enter the width")
area=length*b
print(area)

enter the length 2
enter the width 3


TypeError: can't multiply sequence by non-int of type 'str'

In [45]:
length=float(input("enter the length"))
b=float(input("enter the width"))
area=length*b
print("area is", area)


enter the length 2
enter the width 3


area is 6.0


# exercise 2

In [47]:
item=input("what would you like to buy: ")
price=float(input("what is the price"))
quantity=int(input("how many would you like"))
total =price*quantity
print(f"you have bought {quantity} {item} , your total is {total}")


what would you like to buy:  pizza
what is the price 10.99
how many would you like 3


you have bought 3 pizza , your total is 32.97


# madlibs game = word game where yiu create a story by filling in blanks with random words

In [50]:
adjective1=input("enter an adjective(description)")
noun1=input("enter a noun(person,place , thing)")
adjective2=input("enter an adjective(description)")
verb1=input("enter a verb ending with'ing' ")
adjective3=input("enter an adjective(description)")


print(f"Today i went to a {adjective1} zoo")
print(f"in an exhibit, i saw a {noun1}")
print(f"{noun1} was {adjective2} and {verb1}")
print(f"i was {adjective3}!")

enter an adjective(description) suspicious
enter a noun(person,place , thing) mark
enter an adjective(description) angry
enter a verb ending with'ing'  singing
enter an adjective(description) happy


Today i went to a suspicious zoo
in an exhibit, i saw a mark
mark was angry and singing
i was happy!


 # ARITHMETIC OPERATORS

In [52]:
friends=1
friends=friends*3
print(friends)
friends+=1
print(friends)
friends/=2
print(friends)



3
4
2.0


In [53]:
# exponents

In [54]:
friends**=2
print(friends)

4.0


In [1]:
a=2
a**=3
print(a)

8


In [56]:
b=2
r=b%3    # will give the remainder   
print(r)

2


# built in math functions

In [64]:
# round()
x=3.14
c=round(x)
print(c)


3


In [68]:
#abs()
x=-3.14
print(round(x),abs(x))

-3 3.14


In [69]:
#pow()
a=3
result=pow(a,3)
print(result)

27


In [70]:
#max() min()
a=3
b=-9.0
c=55
print("maximum of a,b,c", max(a,b,c),"minimum of a,b,c", min(a,b,c))

maximum of a,b,c 55 minimum of a,b,c -9.0


# imort math

In [2]:
import math

x=8
print(math.pi)
print(math.e)
result=math.sqrt(x)
print(result)
y=0.9
z=-9.8
print(math.ceil(y))
print(math.ceil(z))
print(math.floor(z))
print(math.floor(y))

3.141592653589793
2.718281828459045
2.8284271247461903
1
-9
-10
0


# exercise

In [3]:
import math
radius=float(input("enyer the radius of a circle:"))
circumference=2*math.pi*radius
print(f"the circumference is :{round(circumference,2)}")

enyer the radius of a circle: 2


the circumference is :12.57


In [77]:
import math
radius=float(input("enter the radius of circle"))
area=math.pi*pow(radius,2)
print(round(area))

enter the radius of circle 2


13


In [43]:
import math
a=int(input("enter side a"))
b=int(input("enter side b"))
c=math.sqrt(pow(a,2)+pow(b,2))
print(round(c,2))

enter side a 1
enter side b 2


2.24


# if= do some code IF  some condition is true ELSE  do something else 

In [44]:
age=int(input("enter your age"))
if age>=18:
    print("you are now signed up")
elif age<0:
    print("born first")
else:
    print("fuck off")
    print("you must be 18+")

enter your age 10


fuck off
you must be 18+


In [83]:
response=input("would you like food(Y/N)")
if response=="Y":
    print("have some food")
else:
    print("no food")


would you like food(Y/N) Y


have some food


In [4]:
name=input("enter your name")
if name=="":
    print("type nigga")
else:
    print(f"hello {name}")

enter your name 


type nigga


In [87]:
# boolean
for_sale=True
if for_sale:
    print("dkjbckdj")
else:
    print("dnsk")

dkjbckdj


# calculator program

In [5]:
operator=input("enter an operator(+,-,/,*)")
num1=int(input("enter first number:"))
num2=int(input("enter second number:"))

if operator=="+":
    print(f"sum= {num1+num2}")
elif operator=="-":
    print(f"sub= {num1-num2}")
elif operator=="*":
    print(f"sub= {num1*num2}")
elif operator=="/":
    print(f"sub= {num1/num2}")
else:
    print("choose valid operator ")

enter an operator(+,-,/,*) +
enter first number: 1
enter second number: 2


sum= 3


# python weight converter

In [ ]:
weight=float(input("enter your weight"))
unit=input("kilograms or pounds(K/L)")

if unit=="K":
    weight=weight*2.205
    unit="Lbs"
elif unit=="L":
    weight=weight/2.205
    unit="kg"
else:
    print(f"{unit} was not valid")

print(f"your weight is {weight}{unit}")

# logical operators= evaluate multiple conditions(or,and,not)
 

In [92]:
temp=25

is_raining=False

if temp>35 or temp<0 or is_raining:
    print("the outdoor event is cancelled")
else:
    print("the outdoor event is still scheduled")

the outdoor event is still scheduled


# conditional expression =A one line shortcut for the if-else statement(ternary operator) Print or a assign one of two values based on conditon

In [93]:
num=5
print("positive"if num>0 else "negative")

positive


In [95]:
n=int(input("enter the number: "))
print("even number" if n%2==0 else"odd number")

enter the number:  66


even number


In [97]:
a=int(input("enter a"))
b=int(input("enter b"))
max_num=a if a>b else b
print(max_num)

enter a 1
enter b 2


2


# strings

In [2]:
# len()
# replace()
#find() = first occurence
#rfind() = last occurence


name=input("enter your name")
print(len(name))
print(f"number of letters in your name=",len(name.replace(" ","")))
result=name.find("s")
print(result)

enter your name simon


5
number of letters in your name= 5
0


In [4]:
a="abcdccdc"
a.count("cdc")


2

In [112]:
# capitalize() = used to capitalize only the first letter
# upper() = used for making all letters capital
# lower() = used for making all letters lower
# isdigit() = return true or false 
#isalpha() =The isalpha() method in Python checks whether all characters in a string are alphabetic. 
            #If there are non-alphabetic characters (e.g., spaces, digits, or symbols) in the string, it will return False.

name=input("enter your name")
print(name.capitalize())
print(name.upper())
print(name.lower())
print(name.isdigit())
print(name.isalpha())

enter your name simon j kurian


Simon j kurian
SIMON J KURIAN
simon j kurian
False
False


In [113]:
name=input("enter your name")

print(name.isalpha())

enter your name simonjkurian


True


In [2]:
#count()
i=input("enter your phone number")
print(i.count("7"))
print(i.count())

enter your phone number 7736530635


AttributeError: 'int' object has no attribute 'count'

# The find() method returns the index of the first occurrence of a substring in a string. If the substring is not found, it returns -1.
# count()Counts occurrences of a substring.

In [117]:
#count()
i=input("enter your phone number")
print(i.count("-"))
print(i.count(""))

enter your phone number 123456789


0
10


In [118]:
print(help(str))

Help on class str in module builtins:

class str(object)
 |  str(object='') -> str
 |  str(bytes_or_buffer[, encoding[, errors]]) -> str
 |
 |  Create a new string object from the given object. If encoding or
 |  errors is specified, then the object must expose a data buffer
 |  that will be decoded using the given encoding and error handler.
 |  Otherwise, returns the result of object.__str__() (if defined)
 |  or repr(object).
 |  encoding defaults to sys.getdefaultencoding().
 |  errors defaults to 'strict'.
 |
 |  Methods defined here:
 |
 |  __add__(self, value, /)
 |      Return self+value.
 |
 |  __contains__(self, key, /)
 |      Return bool(key in self).
 |
 |  __eq__(self, value, /)
 |      Return self==value.
 |
 |  __format__(self, format_spec, /)
 |      Return a formatted version of the string as described by format_spec.
 |
 |  __ge__(self, value, /)
 |      Return self>=value.
 |
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |
 |  __getitem__(

# validate user input exercise
1. username is no more than 12 characters
2. username must not contain spaces
3. username must not contain digits

In [131]:
username=input("enter your username: ")

if len(username)>12:
    print("your username must not be more than 12 characters")
elif not username.find(" ")==-1:
    print(" should not contain spaces")
elif not username.isalpha():
    print("no number")
else:
    print(f"welcome {username}")

enter your username:  simon j 


 should not contain spaces


# string indexing

indexing=accessing elements of a sequence using [  ]
         [start : end : stop]

In [3]:
credit_number="1234-5678-9012-3425"
print(credit_number[1])

2


In [139]:
credit_number="1234-5678-9012-3425"
print(credit_number[0:4]) 
print(credit_number[5:9])
print(credit_number[5:])
print(credit_number.replace(credit_number[:5],"4783-"))

1234
5678
5678-9012-3425
4783-5678-9012-3425


# [ 0 : 4 ]  is same as [ : 4 ]

In [144]:
credit_number="1234-5678-9012-3425"
print(credit_number[-1])
print(credit_number[-5])


print(credit_number[::3])

last_digits=credit_number[-4:]
print(last_digits)

5
-
146-135
3425


In [145]:
#reversing

credit_number="1234-5678-9012-3425"
print(credit_number[::-1])

5243-2109-8765-4321


# format specifiers={value:flags} format a value based on what flage are inserted

In [149]:
price1=3.14159
price2=-647.3737
price3=12.5
# decimal flags
print(f"price 1 is{price1:.2f}")
print(f"price 2 is{price2:.2f}")
print(f"price 3 is{price3:.2f}")
#padding
print(f"price 1 is{price1:10}")
print(f"price 2 is{price2:10}")
print(f"price 3 is{price3:10}")

print(f"price 1 is{price1:010}")
print(f"price 2 is{price2:010}")
print(f"price 3 is{price3:010}")

price 1 is3.14
price 2 is-647.37
price 3 is12.50
price 1 is   3.14159
price 2 is -647.3737
price 3 is      12.5
price 1 is0003.14159
price 2 is-0647.3737
price 3 is00000012.5


# while loop

execute some code WHILE some conditions remain true

In [6]:
name=input("enter your name")
while name=="":
    print("you did not enter your name")
    name=input("enter your name")
print(f"hello {name}")

enter your name 


you did not enter your name


enter your name simon


hello simon


In [2]:
age=int(input("enter your age"))
while age<0:
    print("age cannot be negative")
    age=int(input("enter your age"))
print(f"you are {age} years old")

enter your age -1


age cannot be negative


enter your age -2


age cannot be negative


enter your age 23


you are 23 years old


In [ ]:
num=int(input("enter a number between 1-10"))

while num<1 or num>10:
    print("not valid")
    num=int(input("enter a number between 1-10"))
print(f"your number is {num}")

# python compound interest calculator

In [7]:
principle=0
rate=0
time=0

while principle<=0:
    principle=float(input("enter the principle amount: "))
    if principle<=0:
        print("principle amount cant be less than or equal to 0")
while rate<=0:
    rate=float(input("enter the rate: "))
    if rate<=0:
        print("rate cant be less than or equal to 0")
while time<=0:
    time=int(input("enter the time: "))
    if time<=0:
        print("time cant be less than or equal to 0")
total=principle*pow(1+rate/100,time)
print(f"balance after {time} years : ${total:2f}")

enter the principle amount:  0


principle amount cant be less than or equal to 0


enter the principle amount:  0


principle amount cant be less than or equal to 0


enter the principle amount:  10
enter the rate:  1
enter the time:  1


balance after 1 years : $10.100000


In [8]:
principle=0
rate=0
time=0

while True:
    principle=float(input("enter the principle amount: "))
    if principle<0:
        print("principle amount cant be less than or equal to 0")
    else:
        break
        
while True:
    rate=float(input("enter the rate: "))
    if rate<0:
        print("rate cant be less than or equal to 0")
    else:
        break
while True:
    time=int(input("enter the time: "))
    if time<0:
        print("time cant be less than or equal to 0")
    else:
        break
total=principle*pow(1+rate/100,time)
print(f"balance after {time} years : ${total:2f}")

enter the principle amount:  0
enter the rate:  0
enter the time:  0


balance after 0 years : $0.000000


In [9]:
# while True = means your loop runs infinitely

# for loops = execute a block of code a fixed number of times.You can iterate over a range ,string,sequence,etc

In [10]:
for x in range(1,10):
    print(x)

1
2
3
4
5
6
7
8
9


In [11]:
for x in range(1,11):
    print(x)

1
2
3
4
5
6
7
8
9
10


In [12]:
# it excludes the last number or item
#When to Use range in a for Loop
#The range() function is used when you need to iterate over a sequence of numbers or control the loop with an explicit index.
#When NOT to Use range
#You don’t need range when you’re iterating over an iterable object (e.g., lists, strings, dictionaries) directly. 
#Python allows you to iterate over these objects without explicitly using indices.
#Iterating Over a List or Tuple:

In [3]:
for x in reversed(range(1,11)):
    print(x)

10
9
8
7
6
5
4
3
2
1


In [4]:
for x in range(0,10):
    print(x)

0
1
2
3
4
5
6
7
8
9


 we can also include the step function

In [5]:
for x in range(1,11,3):
    print(x)

1
4
7
10


In [7]:
number=1234567893546
for x in number:
    print(x)

TypeError: 'int' object is not iterable

The error occurs because you are trying to iterate over an integer. The variable number is assigned as an integer (or a numeric literal with dashes, which isn't a valid integer format). Python can't directly iterate over integers.

In [8]:
number = "1234-5678-93546"  # treat the number as a string
for x in number:
    print(x)


1
2
3
4
-
5
6
7
8
-
9
3
5
4
6


# continue

In [9]:
for x in range(1,21):
    if x==13:
        continue
    else:
        print(x)

1
2
3
4
5
6
7
8
9
10
11
12
14
15
16
17
18
19
20


# countdown program

In [8]:
import time

time.sleep(3)
print("times up")

times up


In [12]:
import time

my_time=int(input("enter the time in seconds"))
for x in range(0,my_time):
    print(x)
    time.sleep(1)
print("times up")

enter the time in seconds 9


0
1
2
3
4
5
6
7
8
times up


In [15]:
import time

my_time=int(input("enter the time in seconds"))
for x in reversed(range(0,my_time)):
    seconds=x%60
    minutes=int(x/60)%60
    hours=int(x/3600)
    print(f"{hours:02}:{minutes:02}:{seconds:02}")
    time.sleep(1)
print("times up")

enter the time in seconds 4


00:00:03
00:00:02
00:00:01
00:00:00
times up


# nested loops = A loop within another loop (outer,inner)
#                outer loop

#                           inner loop:

In [17]:
for x in range (1,10):
    print(x)

1
2
3
4
5
6
7
8
9


In [18]:
for x in range (1,10):
    print(x,end="")

123456789

In [20]:
for x in range(3):
    for y in range(1,10):
        print(y,end="")

123456789123456789123456789

In [21]:
for x in range(3):
    for y in range(1,10):
        print(y,end="")
    print()

123456789
123456789
123456789


In [9]:
rows=int(input("enter the no. of rows"))
col=int(input("enter the no. of columns"))
symbol=input("enter any symbol")

for x in range(rows):
    for y in range(col):
        print(symbol,end="")
    print()

enter the no. of rows 3
enter the no. of columns 4
enter any symbol %


%%%%
%%%%
%%%%


In [32]:
n=int(input("enter the number"))

for x in range(1,n+1):
    for y in range(1,x+1):
        print(y,end="")
    print()

enter the number 4


1
12
123
1234


Let’s assume n = 4 as an example input. Now, let’s break down the iterations step by step.

First Loop (for x in range(1, n+1)):
This loop controls the rows. It will run from x = 1 to x = n (which is 4 in this case).

1st Iteration of the Outer Loop (x = 1):
The range for x is from 1 to 4+1 = 5, so the outer loop starts with x = 1.

Now, the inner loop starts:

Inner Loop (for y in range(1, x+1)):

Here x = 1, so the inner loop will run from y = 1 to y = 1 (since x+1 = 2).

1st Iteration of Inner Loop: y = 1

It prints 1 without moving to the next line because of end="".
After the inner loop finishes, print() moves the output to the next line.

Output so far:

Copy code
1
2nd Iteration of the Outer Loop (x = 2):
The outer loop moves to the next iteration with x = 2.

Now, the inner loop starts:

Inner Loop (for y in range(1, x+1)):

Here x = 2, so the inner loop will run from y = 1 to y = 2 (since x+1 = 3).

1st Iteration of Inner Loop: y = 1

It prints 1 without moving to the next line.
2nd Iteration of Inner Loop: y = 2

It prints 2 right after 1 (still on the same line).
After the inner loop finishes, print() moves to the next line.

Output so far:

Copy code
1
12
3rd Iteration of the Outer Loop (x = 3):
The outer loop moves to the next iteration with x = 3.

Now, the inner loop starts:

Inner Loop (for y in range(1, x+1)):

Here x = 3, so the inner loop will run from y = 1 to y = 3 (since x+1 = 4).

1st Iteration of Inner Loop: y = 1

It prints 1.
2nd Iteration of Inner Loop: y = 2

It prints 2 right after 1.
3rd Iteration of Inner Loop: y = 3

It prints 3 right after 2.
After the inner loop finishes, print() moves to the next line.

Output so far:

Copy code
1
12
123
4th Iteration of the Outer Loop (x = 4):
The outer loop moves to the next iteration with x = 4.

Now, the inner loop starts:

Inner Loop (for y in range(1, x+1)):

Here x = 4, so the inner loop will run from y = 1 to y = 4 (since x+1 = 5).

1st Iteration of Inner Loop: y = 1

It prints 1.
2nd Iteration of Inner Loop: y = 2

It prints 2 right after 1.
3rd Iteration of Inner Loop: y = 3

It prints 3 right after 2.
4th Iteration of Inner Loop: y = 4

It prints 4 right after 3.
After the inner loop finishes, print() moves to the next line.

Final Output:

yaml
Copy code
1
12
123
1234
Conclusion:
The outer loop controls how many rows will be printed, starting from x = 1 up to n.
The inner loop controls how many numbers will be printed in each row. It runs from 1 to the current row number x.
Each row prints the numbers from 1 to x, and then the print() moves to a new line.
For every new row, one more number is printed than the previous row.

# collection = single "variable" used to store multiple values
# list =[] ordered and changeable. DUPLICATES OK
# set = {} unordered and immutable, but Add/Remove OK ,NO DUPLICTES
# tuple=() ordered and unchangeable ,duplicates OK, FASTER

In [34]:
# COLLECTION


#list
fruit=["apple","orange","banana","coconut"]
print(fruit)

['apple', 'orange', 'banana', 'coconut']


In [35]:
fruit=["apple","orange","banana","coconut"]
print(fruit[1])

orange


In [36]:
fruit=["apple","orange","banana","coconut"]
print(fruit[4])

IndexError: list index out of range

In [37]:
fruit=["apple","orange","banana","coconut"]
print(fruit[0:3])

['apple', 'orange', 'banana']


In [38]:
fruit=["apple","orange","banana","coconut"]
print(fruit[::2])

['apple', 'banana']


In [39]:
fruit=["apple","orange","banana","coconut"]
for x in fruit:
    print(x)

apple
orange
banana
coconut


In [41]:
fruit=["apple","orange","banana","coconut"]
for x in fruit:
    print(x,end=" ")

apple orange banana coconut 

In [42]:
print(dir)

<built-in function dir>


In [43]:
print(help(fruit))

Help on list object:

class list(object)
 |  list(iterable=(), /)
 |
 |  Built-in mutable sequence.
 |
 |  If no argument is given, the constructor creates a new empty list.
 |  The argument must be an iterable if specified.
 |
 |  Methods defined here:
 |
 |  __add__(self, value, /)
 |      Return self+value.
 |
 |  __contains__(self, key, /)
 |      Return bool(key in self).
 |
 |  __delitem__(self, key, /)
 |      Delete self[key].
 |
 |  __eq__(self, value, /)
 |      Return self==value.
 |
 |  __ge__(self, value, /)
 |      Return self>=value.
 |
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |
 |  __getitem__(self, index, /)
 |      Return self[index].
 |
 |  __gt__(self, value, /)
 |      Return self>value.
 |
 |  __iadd__(self, value, /)
 |      Implement self+=value.
 |
 |  __imul__(self, value, /)
 |      Implement self*=value.
 |
 |  __init__(self, /, *args, **kwargs)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |
 |  __it

In [44]:
fruit=["apple","orange","banana","coconut"]
print(len(fruit))

4


In [45]:
print("apple"in fruit )

True


Some of the methods found in the list

In [46]:
# APPEND
fruit=["apple","orange","banana","coconut"]
fruit.append("chocolatee")
print(fruit)

['apple', 'orange', 'banana', 'coconut', 'chocolatee']


In [50]:

#REMOVE
fruit=["apple","orange","banana","coconut"]
fruit.remove("apple")
print(fruit)

['orange', 'banana', 'coconut']


In [52]:
#INSERT
fruit=["apple","orange","banana","coconut"]
fruit.insert(0,"pinapple")
print(fruit)

['pinapple', 'apple', 'orange', 'banana', 'coconut']


In [54]:
# SORT

fruit.sort()
print(fruit)


['apple', 'banana', 'coconut', 'orange', 'pinapple']


In [55]:
# REVERSE
fruit.reverse()
print(fruit)

['pinapple', 'orange', 'coconut', 'banana', 'apple']


In [56]:
#clear
fruit.clear()
print(fruit)

[]


In [59]:
#INDEX

fruit=["apple","orange","banana","coconut"]
print(fruit.index("coconut")+1)

4


In [60]:
# COUNT
print(fruit.count("banana"))

1


# SET

In [61]:
fruits={"apple","orange","banana","coconut"}
print(dir(fruits))

['__and__', '__class__', '__class_getitem__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__iand__', '__init__', '__init_subclass__', '__ior__', '__isub__', '__iter__', '__ixor__', '__le__', '__len__', '__lt__', '__ne__', '__new__', '__or__', '__rand__', '__reduce__', '__reduce_ex__', '__repr__', '__ror__', '__rsub__', '__rxor__', '__setattr__', '__sizeof__', '__str__', '__sub__', '__subclasshook__', '__xor__', 'add', 'clear', 'copy', 'difference', 'difference_update', 'discard', 'intersection', 'intersection_update', 'isdisjoint', 'issubset', 'issuperset', 'pop', 'remove', 'symmetric_difference', 'symmetric_difference_update', 'union', 'update']


# we cannot use indexing in set 

In [62]:
fruits.add("pinapple")
print(fruits)

{'orange', 'apple', 'banana', 'coconut', 'pinapple'}


# fruits.add("pinapple")
# fruits.remove("apple")
# fruits.pop()
# fruits.clear()

# tuple
# ORDERED ,UNCHANGEABLE, DUPLICATE OK, FASTER

In [63]:
fruit=("apple","orange","banana","coconut")
print(dir(fruit))

['__add__', '__class__', '__class_getitem__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmul__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'count', 'index']


In [64]:
# indexing is possible

# 2D LISTS

In [5]:
fruit=["apple","orange","banana","coconut"]
vegetables=["celery","carrots","potatoes"]
meats=["chicken","fish","turkey"]

groceries=[fruit,vegetables,meats]

print(fruit)

print(groceries)

['apple', 'orange', 'banana', 'coconut']
[['apple', 'orange', 'banana', 'coconut'], ['celery', 'carrots', 'potatoes'], ['chicken', 'fish', 'turkey']]


In [9]:
fruit=["apple","orange","banana","coconut"]
vegetables=["celery","carrots","potatoes"]
meats=["chicken","fish","turkey"]

groceries=[fruit,vegetables,meats]
print(groceries[0])
print(groceries[1])
print(groceries[0][0])   # row 0 column 0


['apple', 'orange', 'banana', 'coconut']
['celery', 'carrots', 'potatoes']
apple


In [10]:
# OR
groceries=[["apple","orange","banana","coconut"],
           ["celery","carrots","potatoes"],
           ["chicken","fish","turkey"]]
print(groceries)

[['apple', 'orange', 'banana', 'coconut'], ['celery', 'carrots', 'potatoes'], ['chicken', 'fish', 'turkey']]


# if you need iterate u can use for loop nested

In [14]:
groceries=[["apple","orange","banana","coconut"],
           ["celery","carrots","potatoes"],
           ["chicken","fish","turkey"]]

for collection in groceries:
    print(collection)

for i in groceries:
    for j in i:
        print(j,end=" ")
    print()

['apple', 'orange', 'banana', 'coconut']
['celery', 'carrots', 'potatoes']
['chicken', 'fish', 'turkey']
apple orange banana coconut 
celery carrots potatoes 
chicken fish turkey 


# we can even use the tupples

In [16]:
num_pad=((1,2,3),
         (4,5,6),
         (7,8,9),
         ("*",0,"#"))
for row in num_pad:
    for num in row:
        print(num,end=" ")
    print()

1 2 3 
4 5 6 
7 8 9 
* 0 # 


# quiz game

In [19]:
questions=("How many elementsare in the periodic table?",
          "which animal lays the largest egg?",
           "how many bones in the human body")

options=(("A.116","B.387","C.67","D.878"),
         ("A.whale","B.crocodile","C.tiger","D.lion"),
         ("A.206","B.207","C.208","D.209"))

answers=("C","D","A")
guesses=[]
score=0
question_num=0

for question in questions:
    print("-----------------")
    print(question)
    for option in options[question_num]:
        print(option)
    guess=input("enter a,b,c,d").upper()
    guesses.append(guess)
    if guess==answers[question_num]:
        score+=1
        print("correct")
    else:
        print("incorrect")
    question_num+=1

print("your score is",score)

-----------------
How many elementsare in the periodic table?
A.116
B.387
C.67
D.878


enter a,b,c,d a


incorrect
-----------------
which animal lays the largest egg?
A.whale
B.crocodile
C.tiger
D.lion


enter a,b,c,d a


incorrect
-----------------
how many bones in the human body
A.206
B.207
C.208
D.209


enter a,b,c,d a


correct
your score is 1


# dictionary = collection of{key:value} pairs
# ordered and unchangeable
# no duplicates

In [20]:
capitals ={"USA":"washingtion",
          "india":"delhi",
          "china":"beijing",
          "russia":"moscow"}
print(capitals)

{'USA': 'washingtion', 'india': 'delhi', 'china': 'beijing', 'russia': 'moscow'}


In [21]:
print(dir(capitals))

['__class__', '__class_getitem__', '__contains__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__ior__', '__iter__', '__le__', '__len__', '__lt__', '__ne__', '__new__', '__or__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__ror__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'clear', 'copy', 'fromkeys', 'get', 'items', 'keys', 'pop', 'popitem', 'setdefault', 'update', 'values']


In [22]:
print(help(capitals))

Help on dict object:

class dict(object)
 |  dict() -> new empty dictionary
 |  dict(mapping) -> new dictionary initialized from a mapping object's
 |      (key, value) pairs
 |  dict(iterable) -> new dictionary initialized as if via:
 |      d = {}
 |      for k, v in iterable:
 |          d[k] = v
 |  dict(**kwargs) -> new dictionary initialized with the name=value pairs
 |      in the keyword argument list.  For example:  dict(one=1, two=2)
 |
 |  Built-in subclasses:
 |      StgDict
 |
 |  Methods defined here:
 |
 |  __contains__(self, key, /)
 |      True if the dictionary has the specified key, else False.
 |
 |  __delitem__(self, key, /)
 |      Delete self[key].
 |
 |  __eq__(self, value, /)
 |      Return self==value.
 |
 |  __ge__(self, value, /)
 |      Return self>=value.
 |
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |
 |  __getitem__(self, key, /)
 |      Return self[key].
 |
 |  __gt__(self, value, /)
 |      Return self>value.
 |
 |  __init

In [23]:
# to get some values from the dictionary



# we will use .get()

In [24]:
capitals ={"USA":"washingtion",
          "india":"delhi",
          "china":"beijing",
          "russia":"moscow"}

print(capitals.get(india))

NameError: name 'india' is not defined

In [11]:
capitals ={"USA":"washingtion",
          "india":"delhi",
          "china":"beijing",
          "russia":"moscow"}
print(capitals.get("russia"))

moscow


# .update() method

In [26]:
capitals.update({"germany":"berlin"})
print(capitals)

{'USA': 'washingtion', 'india': 'delhi', 'china': 'beijing', 'russia': 'moscow', 'germany': 'berlin'}


In [27]:
capitals.update({"USA": "detroit"})
print(capitals)

{'USA': 'detroit', 'india': 'delhi', 'china': 'beijing', 'russia': 'moscow', 'germany': 'berlin'}


# .pop() method

In [28]:
capitals.pop("china")
print(capitals)

{'USA': 'detroit', 'india': 'delhi', 'russia': 'moscow', 'germany': 'berlin'}


In [29]:
# pop item method
# it will remove last element from our dictionary
capitals.popitem()
print(capitals)

{'USA': 'detroit', 'india': 'delhi', 'russia': 'moscow'}


# .clear() method


In [30]:
capitals.clear()
print(capitals)

{}


# .keys method = will get all of the key

In [31]:
keys=capitals.keys()
print(keys)

dict_keys([])


In [32]:
capitals ={"USA":"washingtion",
          "india":"delhi",
          "china":"beijing",
          "russia":"moscow"}
keys=capitals.keys()
print(keys)

dict_keys(['USA', 'india', 'china', 'russia'])


In [33]:
for key in capitals.keys():
    print(key)

USA
india
china
russia


# .values method

In [34]:
for value in capitals.values():
    print(value)

washingtion
delhi
beijing
moscow


# .items method 

In [36]:
items=capitals.items()
print(items)

dict_items([('USA', 'washingtion'), ('india', 'delhi'), ('china', 'beijing'), ('russia', 'moscow')])


In [37]:
for key,value in capitals.items():
    print(f"{key}:{value}")

USA:washingtion
india:delhi
china:beijing
russia:moscow


# consession stand program

In [18]:
menu={"pizza":3,
      "nachos":4.5,
      "popcorn":6,
      "fries":2.50,
      "chips":1.00,
      "pretzel":3.50,
      "soda":3.00,
      "lemonade":4.25}
cart=[]
total=0

print("-----------------menu----------------")
for key,value in menu.items():
    print(f"{key:10}: ${value:.2f}")

print("--------------------------------------")

while True:
    food=input("select an item (q to quit): ").lower()
    if food=="q":
        break
    elif menu.get(food) is not None:
        cart.append(food)
print("-----------------your order----------------")
for food in cart:
    total=total+menu.get(food)
    print(food,end=" ")
print()
print(f"total is : ${total:.2f}")


-----------------menu----------------
pizza     : $3.00
nachos    : $4.50
popcorn   : $6.00
fries     : $2.50
chips     : $1.00
pretzel   : $3.50
soda      : $3.00
lemonade  : $4.25
--------------------------------------


select an item (q to quit):  popcorn
select an item (q to quit):  soda
select an item (q to quit):  q


-----------------your order----------------
popcorn soda 
total is : $9.00


# generating random number 

In [20]:
import random
number =random.randint(1,6)
print(number)

2


In [22]:
import random
#.randint()
low=1
high=100
number =random.randint(low,high)
print(number)

93


In [26]:
#.choice()
import random 
options=("rock","paper","scissors")
option=random.choice(options)
print(option)

rock


In [27]:
import random
cards=["2","3","4","5","6"]
shuf=random.shuffle(cards)
print(shuf)

None


In [30]:
import random
cards=["2","3","4","5","6"]
random.shuffle(cards)
print(cards)

['6', '3', '5', '2', '4']


# number guessing game

In [32]:
import random

low_num=1
high_num=100

answer=random.randint(low_num,high_num)
guesses=0
is_running=True

print("python number guessing game ")
print(f"select a number between {low_num} and {high_num}")

while is_running:
    guess=int(input("enter your guess"))
    if guess.isdigit():
        guesses+=1

        if guess<low_num or guess>high_num:
            print("number out of range")
            print(f"please select a number between 1 and 100")
        elif guess<answer:
            print("two low! try again")
        elif guess>answer:
            print("two high try again")
        else:
            print(f"correct number of guesses: {guesses}")
    else:
        print("invalid")
        print(f"please select a number between 1 and 100")

python number guessing game 
select a number between 1 and 100


enter your guess 3


AttributeError: 'int' object has no attribute 'isdigit'

python number guessing game 
select a number between 1 and 100


enter your guess 3


two low! try again


enter your guess 10


two low! try again


enter your guess 90


two high try again


enter your guess 85


two high try again


enter your guess 65


two high try again


enter your guess 45


two high try again


enter your guess 20


two low! try again


enter your guess 30


two low! try again


enter your guess 35


two high try again


enter your guess 32


two low! try again


enter your guess 31


two low! try again


enter your guess 34


two high try again


enter your guess 33


correct number of guesses: 13


enter your guess ld,


invalid
please select a number between 1 and 100


enter your guess d


invalid
please select a number between 1 and 100


enter your guess 2


two low! try again


enter your guess 2


two low! try again


enter your guess 2


two low! try again


enter your guess 2


two low! try again


enter your guess 2


two low! try again


enter your guess 2


two low! try again


enter your guess 2


two low! try again


enter your guess 2


two low! try again


enter your guess 2


two low! try again


enter your guess 2


two low! try again


enter your guess 2


two low! try again


enter your guess 2


two low! try again


KeyboardInterrupt: Interrupted by user

In [36]:
import random

low_num=1
high_num=100

answer=random.randint(low_num,high_num)
guesses=0
is_running=True

print("python number guessing game ")
print(f"select a number between {low_num} and {high_num}")

while is_running:
    guess=input("enter your guess")
    if guess.isdigit():
        guess = int(guess)
        guesses+=1

        if guess<low_num or guess>high_num:
            print("number out of range")
            print(f"please select a number between 1 and 100")
        elif guess<answer:
            print("two low! try again")
        elif guess>answer:
            print("two high try again")
        else:
            print(f"correct number of guesses: {guesses}")
            break
    else:
        print("invalid")
        break
        print(f"please select a number between 1 and 100")
     

python number guessing game 
select a number between 1 and 100


enter your guess s


invalid
please select a number between 1 and 100


# rock paper scissors

In [38]:
import random

options=("rock","paper","scissors")
player=None
computer=random.choice(options)

running =True
while running:
    while player not in options:
        player = input("enter choice rock paper scissors: ")
    
    print(f"player: {player}")
    print(f"computer:{computer}")
    
    if player==computer:
        print("its a tie")
    elif player=="rock"and computer=="scissors":
        print("you win")
    elif player=="paper"and computer=="rock":
        print("you win!")
    elif player=="scissors" and computer=="paper":
        print("you win!")
    else:
        print("you lose")
    play_again=input("play again y/n").lower()
    if not play_again=="y":
        running=False
print("thanks for playing")



enter choice rock paper scissors:  rock


player: rock
computer:scissors
you win


# function = a block of reusable code place() after the function name to invoke it 

In [40]:
def happy():
    print("hello")
    print("world")

happy()
happy()

hello
world
hello
world


In [41]:
# position of the parameters or the arguments passed also matters

In [2]:
def display_invoice(username,amount,due_date):
    print(f"hello {username}")
    print(f"your bill of ${amount} is due {due_date}")

display_invoice("simon",100,"01/01")

hello simon
your bill of $100 is due 01/01


# return = statement used to end a function and send a result back to the caller

In [5]:
def add(x,y):
    z=x+y
    return z

def subtract(x,y):
     z=x-y
     return z

print(add(1,2))
print(add(1,2))
    

3
3


In [7]:
def create_name(first,last):
    first=first.capitalize()
    last=last.capitalize()
    return first+" "+last

full_name=create_name("simon","kurian")
print(full_name)

Simon Kurian


# default arguments= A default value for certain parameters
#                    default is is used when that argument is ommited
#                    make your functions more flexible ,reduces # of arguments
#                    1.positional 2.default 3.keyword 4.arbitary

In [8]:
def net_price(list_price,discount,tax):
    return list_price*(1-discount)*(1+tax)

print(net_price(500,0,0.05))

525.0


In [9]:
def net_price(list_price,discount=0,tax=0.05):
    return list_price*(1-discount)*(1+tax)

print(net_price(500))

525.0


# program

In [10]:
import time

def count(start,end):
    for x in range(start,end+1):
        print(x)
        time.sleep(1)
    print("done")

count(0,10)

0
1
2
3
4
5
6
7
8
9
10
done


In [11]:
# if we are using default arguments it should come after non default argument

# default arguments = an argument preceeded by an identifier, helps with readability, order of arguments don't matter


In [14]:
def hello(greeting,title,first,last):
    print(f"{greeting} {title}.{first}{last}")

hello("helllo","mr","simon","kurian")

helllo mr.simonkurian


In [15]:
def hello(greeting,title,first,last):
    print(f"{greeting} {title}.{first}{last}")

hello("helllo",title="mr",last="simon",first="kurian")

helllo mr.kuriansimon


In [16]:
def hello(greeting,title,first,last):
    print(f"{greeting} {title}.{first}{last}")

hello(title="mr",last="simon",first="kurian")

TypeError: hello() missing 1 required positional argument: 'greeting'

In [17]:
for x in range(1,11):
    print(x, end=" ")
    

1 2 3 4 5 6 7 8 9 10 

In [18]:
print("1","2",sep="-")

1-2


# *args= allows yoou to pass multiple non key arguments
# **kwargs= allows yoou to pass multiple keyword -arguments
# *unpacking operator

In [19]:
def add(*args):
    return a+b

add(add(1,2))

NameError: name 'a' is not defined

In [20]:
def add(*args):
    print(type(args))

add(1,2)

<class 'tuple'>


In [21]:
def add(*args):
    total=0
    for arg in args:
        total+=arg
    return total

print(add(1,2,3,4))

10


In [22]:
# parameter name can vary

In [25]:
def print_address(**kwargs):
    for key,value in kwargs.items():
        print(f"{key}: {value}")

print_address(street="jfbjfbdj",
             city="hbdshbd",
             state="india",
             )

street: jfbjfbdj
city: hbdshbd
state: india


# PROBLEMS

In [1]:
def swap_case(s):
    s = list(s)  # Convert string to a list
    for i in range(len(s)):
        if s[i].islower():  # Check if the character is lowercase
            s[i] = s[i].upper()  # Convert to uppercase
        elif s[i].isupper():  # Check if the character is uppercase
            s[i] = s[i].lower()  # Convert to lowercase
    return ''.join(s)  # Join the modified list back into a string

if __name__ == '__main__':
    s = input("Enter a string: ")
    result = swap_case(s)
    print(result)


Enter a string:  Www.HackerRank.com


wWW.hACKERrANK.COM


In [3]:
def split_and_join(line):
    
    line=line.replace(" ","-")
    return line

if __name__ == '__main__':
    line = input()
    result = split_and_join(line)
    print(result)

 this is a string


this-is-a-string


In [3]:
if __name__ == '__main__':
    n = int(input())
    arr = map(int, input().split())
    for i in arr:
        print(i,end="")

 3
 1 2 3 4 5 5 6 7 


12345567

In [ ]:
if __name__ == '__main__':
    n = int(input())
    arr = map(int, input().split())
    u = list(set(arr))
    u.sort(reverse=True)
    print(u[1])

In [8]:
if __name__ == '__main__':
    list1=[]
    for _ in range(int(input())):
        name = input()
        score = float(input())
        list1.append([name,score])
    smallest=min([score for _,score in list1])
    second_smallest=min([score for _,score in list1 if score!=smallest])
    second_smallest_name=[name for name,score in list1 if score==second_smallest]
    second_smallest_name.sort()
    for name in second_smallest_name:
        print(name)
      

SyntaxError: did you forget parentheses around the comprehension target? (14049481.py, line 7)

In [ ]:
if __name__ == '__main__':
    n = int(input())
    student_marks = {}
    for _ in range(n):
        name, *line = input().split()
        scores = list(map(float, line))
        student_marks[name] = scores
    query_name = input()
    if query_name in student_marks:
        scores = student_marks[query_name]
        avg_score = sum(scores) / len(scores)
       
        print(f"{avg_score:.2f}")
        

# object oriented programming


In [2]:
class Car:
    def __init__(self,model,year,color,for_sale): # it is a constructor method to create objects its a dunder method =double underscore
        self.model=model
        self.year=year
        self.color=color
        self.for_sale=for_sale

car1=Car("mustang",2024,"red",False)
print(car1)
       
        

In [3]:
# the output we get is he memory address of this car object where it is located

In [7]:
class Car:
    def __init__(self,model,year,color,for_sale): 
        self.model=model
        self.year=year
        self.color=color
        self.for_sale=for_sale

car1=Car("mustang",2024,"red",False)
print(car1.model)
print(car1.color)
print(car1.for_sale)
       
car2=Car("bmw",2024,"blue",True)
print(car2.model)

mustang
red
False
bmw


In [8]:
# METHODS are actions that an object can perform

In [13]:
class Car:
    def __init__(self,model,year,color,for_sale): 
        self.model=model
        self.year=year
        self.color=color
        self.for_sale=for_sale
    def drive(self):
        print(f"you drive the {self.model}")
    def stop(self):
        print("you stop the car")

car1=Car("mustang",2024,"red",False)
car1.drive()
car1.stop()

you drive the mustang
you stop the car


# class variables= shared among all instances of a class defined outside the construtor, allow you to share dat aong all objects created from that class

In [14]:
class Student:
    def __init__(self,name,age):
        self.name=name
        self.age=age

student1=Student("simmon",20)
student2=Student("rohan",19)
print(student1.age)

20


In [17]:
class Student:

    class_year=2024
    def __init__(self,name,age):
        self.name=name
        self.age=age

student1=Student("simmon",20)
student2=Student("rohan",19)
print(student1.class_year)
print(student2.class_year)
print(Student.class_year)

2024
2024
2024
